bin/docker-build.js

```diff
@@ -453,25 +471,183 @@ function buildFromContext(opts, callback) {
+        var id;
+        var cbEvent;
+        try {
+            event = JSON.parse(event);
+        } catch (e) {
+            log.error('Build: invalid json: %s - ignoring', event);
             return;
         }
-        var snapNames = allSnaps.map(function (s) { return s.name; }).join(',');
-        var zfsSnaps = sprintf('zones/%s@%s', builder.zoneUuid, snapNames);
-        builder.log.debug('Destroying snapshots %j', zfsSnaps);
-        zfs.destroy(zfsSnaps, callback);
+        switch (event.type) {
+            case 'callback':
+                id = event.messageId;
+                cbEvent = pendingCallbackEvents[id];
+                assert.object(cbEvent, 'cbEvent with messageId ' + id);
+                delete pendingCallbackEvents[id];
+                if (event.error) {
+                    cbEvent.callback(new Error(event.error));
+                } else {
+                    cbEvent.callback(null, event.result);
+                }
+                break;
+            default:
+                log.error('Unhandled socket event - ignoring: %j', event);
+                break;
+        }
+    });
+
+    var buildOpts = {
+        commandType: 'commit',
+        containerRootDir: path.join('/zones', opts.uuid, 'root'),
+        contextFilepath: '<none for docker commit>',  // XXX: Fake context.
+        log: opts.log,
+        nocache: true,
+        uuid: opts.uuid,
+        workDir: path.join('/zones', opts.uuid, 'config')
+    };
+    var builder = new dockerbuild.Builder(buildOpts);
+
+    // Add the base image as the first snapshot.
+    var fullSnapshotName = sprintf('zones/%s@final', fromImageUuid);
+    log.debug('Base image snapshot is %j', fullSnapshotName);
+    gSnapshots.push({
+        name: fullSnapshotName,
+        layerIdx: 0
     });
+
+    builder.on('message', function (event) {
+        switch (event.type) {
+            case 'message':
+            case 'stdout':
+                sendEvent(event, {socket: socket, log: log});
+                break;
+            default:
+                log.error('Unhandled builder event - ignoring: %j', event);
+                break;
+        }
+    });
+
+    builder.on('task', function (event) {
+        switch (event.type) {
+            case 'image_create':
+                sendEvent(event, {socket: socket, log: log});
+                break;
+            case 'commit_finished':
+                sendEvent(event, {socket: socket, log: log});
+                break;
+            default:
+                log.error('Unhandled commit task - ignoring: %j', event);
+                if (event.callback) {
+                    event.callback(new Error('Unhandled build task: '
+                                            + event.type));
+                    return;
+                }
+                break;
+        }
+
+        if (event.callback) {
+            pendingCallbackEvents[event.messageId] = event;
+        }
+    });
+
+    builder.on('end', function (err) {
```

Comment:
Does `dockerbuild.Builder` instances emit `'error'` events, and if so should we add a listener for these?